### ISSUE TITLE
vector Gaussian likelihood variance for varying measurement uncertainty.

### ISSUE THREAD

**Issue by Joshuaalbert** [GPflow/GPflow#550]
I would be very useful to have a vector variance in the Gaussian likelihood, as it is very often the case that measurements do not have a constant uncertainty.

**Comment by hughsalimbeni** [comment]
If the data is NxD, do you mean constant between dimension (D) or constant between data points (N)?

**Comment by Joshuaalbert** [comment]
I mean the sigma_n^2 in (2.20) in http://www.gaussianprocess.org/gpml/chapters/RW.pdf. This is the measurement uncertainty, and not a proper hyper parameter. More generally sigma_n^2 * Eye should be Diag(sigma_n^2). This can be achieved in a hacky way because of the broadcasting rules of tensorflow via:
```
sigma_n = [...]
likelihood = GPflow.likelihoods.Gaussian(sigma_n[:,None])
#if you don't want to treat the likelihood variance as a hyperparameter 
likelihood.variance.set_trainable(False) 
```

**Comment by Joshuaalbert** [comment]
## Hacky solution

In the case where you have measurements with well defined measurement uncertainty, where you wish to use a GP with Gaussian likelihood to interpolate or smooth you can use,
```
sigma_n = [...] #f_i error bars
likelihood = GPflow.likelihoods.Gaussian(sigma_n[:,None]**2)
likelihood.variance.set_trainable(False)
m = GPflow.models.GPR(X, Y, kern=kern, mean_function=mean)
GPflow.models.GPModel.__init__(m,GPflow.DataHolder(X),GPflow.DataHolder(Y),
                           kern,likelihood,mean)
```

**Comment by hughsalimbeni** [comment]
That will work for `GPR`, but watch out that `SGPR` won't work with this approach. If you're using one of the variational models then you can also do this as long as you don't ask for full covariance predictions.

**Comment by awav** [comment]
@Joshuaalbert, do not called private functions, especially `__init__`, you never know what happens in the background. Define another class:

```
class GPR2(gpflow.models.GPR):
     def __init__(self, likelihood, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.likelihood = likelihood
```

or

```
with gpflow.defer_build():
    sigma_n = [...] #f_i error bars
    likelihood = GPflow.likelihoods.Gaussian(sigma_n[:,None]**2)
    likelihood.variance.set_trainable(False)
    m = GPflow.models.GPR(X, Y, kern=kern, mean_function=mean)
    m.likelihood = likelihood
    m.compile()
```

**Comment by Joshuaalbert** [comment]
@awav Thanks for the tip. I didn't know of the defer_build. That is how one sets priors in the new API I'm guessing. I just came across gpflow a few days ago. I wrote my own TF-GP code but decided to look out there for better, more full, implementations. I am under the impression that gpflow is in the middle of an API transition, but the codebase seems solid.

## More to the point
Shouldn't the badness behind my hacky way suggest that a control step for a vector measurement uncertainty be included by default? This seems simpler than creating redundant classes.

**Comment by markvdw** [comment]
We're literally just releasing the new version of GPflow, so I guess one could say it is in a transition :) I do really believe the codebase is very solid. The old version was rock solid for what it did.

The new version is much more flexible and I have been using it for my daily research for weeks while it was being developed. The code in master is solid, but I think community-wise we still need to get people up to speed with how to use it, and what has changed. So from that point of view, we're in a bit of a transition.

**Comment by Joshuaalbert** [comment]
Curious, will this request get addressed or will it require writting our own classes?

**Comment by markvdw** [comment]
It should be very easy to adapt the Gaussian likelihood to cater for observations for different measurement noise. However, I don't think there is a plan from within the core developers to implement this right now. However, if you need this feature, we would encourage you to implement it and submit a pull request.

GPflow is an open source project, so we do welcome additions which fit within the codebase and are within the scope of the project (see the [contribution guidelines](https://github.com/GPflow/GPflow/blob/master/contributing.md)). Many of the developers and contributors of GPflow are generally full time researchers themselves. The motivation of developing this package is to share and leverage code advances, to speed up research. As such, we submit code that we develop that is generally useful. However, we can not accommodate every feature request.

**Comment by Joshuaalbert** [comment]
@markvdw thanks, I would be happy to create the feature and make a pull request :)

**Comment by Joshuaalbert** [comment]
I'm going to tackle this now. Ideally, I'd like to make it an option for the Gaussian likelihood (and therefore also hardcoded for the GPR). If so, then I'd make it a kwarg:
```
sigma_Y : (optional) strictly positive float or array_like, or None
    If not None (default), gives the fixed measurement uncertainty for the observations. 
    If array_like then it must be the same shape as observable ``Y``.
    If None, then likelihood variance parameter will be a trainable float.
``` 

Secondly, I would also like to add an option to allow scaling of `sigma_Y`.
```
scale_sigma_Y : (optional) bool
    If True and ``sigma_Y`` is not None then include a trainable multiplicative scalar factor to ``sigma_Y``
    which can be useful when only relative measurement uncertainties are known.
    If ``sigma_Y`` is a scalar and ``scale_sigma_Y`` is True then this the same as leaving ``sigma_Y`` as None.
```

**Comment by markvdw** [comment]
@Joshuaalbert, this looks like a useful contribution. We may need to adapt the exact interface a bit in the discussion of the PR, but overall this seems sensible.

The only thing I would suggest is to implement `scale_sigma_Y` slightly differently. I think this would be best implemented as a transform. Essentially what you seem to be saying, is that you want to constrain the variances to have the same relative values. This constraint can be replicated in a transform which has as parameters the relative scales, and then transforms a single value to an array of the desired size. What do you think?

**Comment by Joshuaalbert** [comment]
Hey thanks! The motivation for the `scale_sigma_Y` is that an observable might be drawn from the distribution `Y~G(Y_obs, (scale_sigma_Y*sigma_y)^2)` and that `scale_sigma_Y` is unknown. An example is when one derives the measurement uncertainties from the measurement uncertainties of another observable. One might then get useful measurement uncertainties modulo some scale factor, which can be treated as a latent model hyper-parameter to be optimized over.

In this case, given that the relative scale is unknown, I'm not sure how to represent it as a transform. Can a transform contain trainable parameters?

**Comment by markvdw** [comment]
My suggestion would be that the transform (say, `ConstrainRatios`) would contain a variable as a property which is the relative variances. The forward transform would multiply a single number by the relative variances. The result would be exactly what you're looking for -- a vector which always has the same relative values.

**Comment by Joshuaalbert** [comment]
I suppose that could be a better way. Then the `sigma_Y` would actually be in the form of a transform and the `scale_sigma_Y` would be the current variance parameter. So it makes it slightly more uniform when the user wants current behaviour (`sigma_Y` means they just don't place the extra transform on the variance). Then `scale_sigma_Y` takes the role of variance. So the equivalent behaviour for `scale_sigma_Y = None` would be add `ConstrainRatios` and to `variance.set_trainable(False)`.

**Comment by Joshuaalbert** [comment]
@markvdw Apologies for taking a while to get to this. I'm running into an issue with your recommended method (See Gaussian2 in gist). Although using two params (a trainable one for absolute scaling, and non trainable one for setting relative errors) works fine (Gaussian1).

MVCE: https://gist.github.com/Joshuaalbert/431e084cb2cd47da723a4436bc7fceb9

**Comment by markvdw** [comment]
After a quick glance at the code: Are you making provisions that the transform will have a smaller free size than transformed size? Essentially, the transform is not bijective, and it maps from R to R^D.

**Comment by Joshuaalbert** [comment]
Thanks for the quick reply. The code is runnable by copy and paste if you wanna give it a go. Essentially if the likelihood param `var` is size > 1 then I use the following non-bijective transform (because I assume broadcasting functions):
```
self.variance = Parameter(
            1., transform = Chain(positive, RelativeUncertainties(var)), dtype=settings.float_type)
```

The RealtiveUncertainties transform is defined as a vector element-wise scaling,
```
class RelativeUncertainties(Transform):
    """
    The transform that will enable vector observational uncertainty
    by forcing the likelihood variance parameter to a vector with
    values proportional to these given relative errors.

    Note: this is used for incorperating vector uncertainties in likelihoods
    by placing this transform on the variance parameter.
    
    Essentially, this is a vector rescaling function.
    """
    def __init__(self, relative_errors = 1.):
        self.relative_errors = np.atleast_1d(relative_errors)

    def forward_tensor(self, x):
        return x * self.relative_errors

    def forward(self, x):
        return x * self.relative_errors

    def backward_tensor(self, y):
        return y / self.relative_errors

    def backward(self, y):
        return y / self.relative_errors

    def log_jacobian_tensor(self, x):
        errors = tf.cast(self.relative_errors, dtype=settings.float_type)
        log_errors = tf.log(errors)
        return tf.reduce_sum(log_errors)

    def __str__(self):
        return "{}*".format(self.relative_errors)
```

**Comment by markvdw** [comment]
Isn't a vector wise rescaling exactly the wrong thing? You want the free state of the transform (i.e. what is transformed by `forward()` to the actual parameter value) to be a single number, right?

Shouldn't 
```
    def forward(self, x):
        return x[0] * self.relative_errors
```
be more what you're looking for?

**Comment by Joshuaalbert** [comment]
The goal is to achieve using the current variance parameter and a tranform,
```
var = 1.0 (trainable, and positive)
relative_variance = [...] (fixed, positive)
variance = var*relative_variance
```
So I'd say the RelativeUncertainty (RU) forward transform should take positive scalar and turn it into a vector that is proportional to the given init vector with proportionality given by the positive scalar. Therefore, as I understand it the transform `RU` forward pass acts on the scalar by performing this transform: `x * relative_variance`. Now say that we have a transformed vector `y` and want to go back to the scalar uncontrained parameter, then I suppose one way is to do `tf.reduce_mean(y / relative_variance)`. Am I missing something here?

**Comment by Joshuaalbert** [comment]
I am working on this topic, however, I've mostly switched to working on implementing with SVGP with Gaussian likelihood. Using transforms doesn't work. The observational variance must be passed just like X or Y, as a dataholder and the Model methods need to be accordingly modified to take the vector observational variance into account. Unfortunately, this results in a code that is not compatible with the generic gpflow Model framework (as things get overloaded). If you have a use case, I can share.

**Comment by chenmaosi** [comment]
I'm working on a project requiring me to smooth a long sequence (2000+ points) with dynamic uncertainty. I tried the `sklearn.gaussian_process`, it supports array-like uncertainty as an input but it solves really slowly. I like gpflow because of its easy interface and the support of GPU (via TensorFlow). I'm new to SVGP or VGP (still reading the papers), if you think they could work for my case, I would like to try your code. Thank you!

**Comment by Joshuaalbert** [comment]
Sorry for the slow reply. See my answer to #619. I suggest you to read [Hensman et al. 2015](http://proceedings.mlr.press/v38/hensman15.pdf) to get up to speed on the SVGP model. However if you only have < 10000 datapoints you can for sure fit that in memory for a decent computer. You'll have to wait some time also for it to plug away but it'll be fine. If you need to do smoothing frequently on large arrays then perhaps you can turn to sparse methods.

To recap my post in #619, you do not want to change the likelihood function itself to weight your data points. Rather you need to modify the `_build_likelihood` method of the `Model` class you are using. In that post I give an example for `GPR`.

**Comment by Joshuaalbert** [comment]
I'm going to close this, as the previous post explains the solution.

**Comment by chenmaosi** [comment]
@Joshuaalbert, Thank you!

**Comment by jens-rataczak** [comment]
Hello, is the explanation in #619 what should be used for a Gaussian likelihood with a vector-input variance (for datapoints with known measurement uncertainites)? I have tried implementing some of these things in the current version of GPFlow without any luck
